sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -1271,42 +1295,45 @@ def task_worker(self):
 
                         # TODO: Use fine grained time when the CMIS memory map is available
                         self.port_dict[lport]['cmis_expired'] = now + datetime.timedelta(seconds=self.CMIS_DEF_EXPIRED)
-                        self.port_dict[lport]['cmis_state'] = self.CMIS_STATE_AP_CONF
-                    elif state == self.CMIS_STATE_AP_CONF:
+                        self.port_dict[lport]['cmis_state'] = self.CMIS_STATE_DP_INIT
+                    elif state == self.CMIS_STATE_DP_INIT:
                         if not self.test_config_error(api, host_lanes, ['ConfigSuccess']):
                             if (expired is not None) and (expired <= now):
                                 self.log_notice("{}: timeout for 'ConfigSuccess'".format(lport))
                                 self.reset_cmis_init(lport, retries + 1)
                             continue
 
+                        # Ensure the Datapath is NOT Activated unless the host Tx siganl is good.
+                        # NOTE: Some CMIS compliant modules may have 'auto-squelch' feature where
+                        # the module won't take datapaths to Activated state if host tries to enable
+                        # the datapaths while there is no good Tx signal from the host-side.
+                        if not is_host_tx_ready(lport):
```

Comment:
Can this be implemented as a event handler? 
1. Subscribe to PORT_TABLE in state-db, then filter to host_tx_ready field. 
2. Based on the host_tx_ready status, set the appropriate CMIS module state will be set here in the event handler.
3. Since the main CMIS task worker keep checking for CMIS state change for all ports, once the point (2) is set, the CMIS task worker will work on the appropriate state.
4. we can introduce a new state between CMIS_STATE_AP_CONF and CMIS_STATE_DP_INIT,  say 'CMIS_STATE_WAIT_FOR_HOST' . so during module insertion the CMIS task worker will set to this state once it completes CMIS_STATE_AP_CONF and this state will be changed to CMIS_STATE_DP_INIT only in the host_tx_ready handler. this way it will be synchornized.
  
Keeping the host_tx_ready processing as event handler will have following benefit.
1. Keeping the host_tx_ready out of CMIS task worker to event handler will benefit other speed to follow the deterministic link bringup approach.
2. It will avoid continuous polling of host-tx-ready using Redis-DB query, which will save time when scaled to fully populated 48/72 ports.

Reply:
@jaganbal-a I think event handling may NOT work when Xcvrd restarts. Consider this:-
1. 400G Port link is up.
2. Xcvrd crashes and in the meantime someone plugs in another 400G module
3. Xcvrd won't get another 'host_tx_ready' event because 'host_tx_ready' doesn't depend upon whether module is present or not. So, Xcvrd will keep waiting for 'host_tx_ready' or unless user does admin shut followed by admin up.

Another similar case as above
1. Xcvrd is running and got 'host_tx_ready'  but module is absent 
2. Xcvrd crashes and comes up
3. Module is plugged in but Xcvrd is not going to receive 'host_tx_ready' and link won't come up

Reply:
@prince 
It is captured as part of design that the host_tx_ready will be updated regardless of transceiver presence, and it is xcvrd job to check the host_tx_ready during module Init during insertion <a local host_tx_ready dictionary will help here>, so the DP state will be moved to DP_INIT/DP_ACTIVATE state based on host_tx_ready.

Reply:
@jaganbal-a Xcvrd process don't have a persistent storage to capture the 'host_tx_ready' state, so it depends upon the Redis-DB i.e it needs to read the 'host_tx_ready' state from the DB. Let Xcvrd read the state when it requires rather than being notified. Because if we have to take the approach of notification, then we need to handle the above Xcvrd exit case separately by reading the DB which can be avoided if Xcvrd can read the DB when required.

Reply:
@prgeor 

>  Let Xcvrd read the state when it requires rather than being notified. 

Can you clarify as to what is meant by 'when it requires'? 
Does it mean continuous polling of host-tx-ready using Redis-DB query OR what?
When interface/port moves from config shut to no shut or vice-versa, won't it a notification of host_tx_ready from redisDB otherwise how would it xcvrd be notified of this? repeated/continuous polling indefinitely without any event happening in the system is waste of cpu cycles / system resources. 
Also config interface shut/no shut are rare events along with optics removal/insertion
Also, per HLD we decided to go with notification based mechanism of host_tx_ready.
To handle xcvrd restart case, as Jagan mentioned above, better to check the host_tx_ready during module Init during insertion as source of truth.

Reply:
@shyam77git @jaganbal-a please take a look again as i updated the code.